MSAL/test/integration/native_auth/end_to_end/mfa/MSALNativeAuthSignInWithMFAEndToEndTests.swift

```diff
@@ -0,0 +1,226 @@
+// THE SOFTWARE.  
+
+import Foundation
+import XCTest
+import MSAL
+
+final class MSALNativeAuthSignInWithMFAEndToEndTests: MSALNativeAuthEndToEndPasswordTestCase {
+
+    func test_signInUsingPasswordWithMFASubmitWrongChallengeResendChallengeThen_completeSuccessfully() async throws {
+        throw XCTSkip("Missing username for MFA user")
+        guard let username = retrieveUsernameForSignInUsernamePasswordAndMFA(), 
+                let password = await retrievePasswordForSignInUsername(),
+                let awaitingMFAState = await signInUsernameAndPassword(username: username, password: password)
+        else {
+            XCTFail("Something went wrong")
+            return
+        }
+        
+        // Request to send challenge to the default strong auth method
+        let mfaExpectation = expectation(description: "mfa")
+        let mfaDelegateSpy = MFARequestChallengeDelegateSpy(expectation: mfaExpectation)
+        
+        awaitingMFAState.requestChallenge(delegate: mfaDelegateSpy)
+        
+        await fulfillment(of: [mfaExpectation])
+        
+        guard mfaDelegateSpy.onVerificationRequiredCalled, let mfaRequiredState = mfaDelegateSpy.newStateMFARequired else {
+            XCTFail("Challenge not sent to MFA method")
+            return
+        }
+        
+        // Now submit the wrong email OTP code
+        let submitWrongChallengeExpectation = expectation(description: "submitChallenge")
+        let mfaSubmitWrongChallengeDelegateSpy = MFASubmitChallengeDelegateSpy(expectation: submitWrongChallengeExpectation)
+
+        mfaRequiredState.submitChallenge(challenge: "000000", delegate: mfaSubmitWrongChallengeDelegateSpy)
+
+        await fulfillment(of: [submitWrongChallengeExpectation])
+
+        XCTAssertTrue(mfaSubmitWrongChallengeDelegateSpy.onMFASubmitChallengeErrorCalled)
+        XCTAssertEqual(mfaSubmitWrongChallengeDelegateSpy.error?.isInvalidChallenge, true)
+        
+        guard let mfaRequiredState = mfaSubmitWrongChallengeDelegateSpy.newStateMFARequiredState else {
+            XCTFail("New state not received after SDK error")
+            return
+        }
+        
+        // Resend code to default auth method
+        
+        let mfaResendChallengeExpectation = expectation(description: "mfa")
+        let mfaResendChallengeDelegateSpy = MFARequestChallengeDelegateSpy(expectation: mfaResendChallengeExpectation)
+        
+        mfaRequiredState.requestChallenge(delegate: mfaResendChallengeDelegateSpy)
+        
+        await fulfillment(of: [mfaResendChallengeExpectation])
+        
+        guard mfaResendChallengeDelegateSpy.onVerificationRequiredCalled, let mfaRequiredState = mfaDelegateSpy.newStateMFARequired else {
+            XCTFail("Challenge not sent to MFA method")
+            return
+        }
+
+        // Now retrieve and submit the email OTP code
+        await retrieveAndSubmitCode(state: mfaRequiredState, username: username)
```

Comment:
Should you not check that now the state is `Complete`? Or is that done inside `retrieveAndSubmitCode`?

Reply:
I see that's done in `retrieveAndSubmitCode`. IMO that method name is incorrect, as it does more than just submit the code; it also asserts a few things. But nit :)

Reply:
yeah, it's not clear. I'll rename it to something like `completeAuthentication`